Use SplitN rather than Split on os.Environ return value.

smoser opened the pull request:
Environment variables can contain a '=' in their value
   KEY=BAR=1 sh -c 'echo KEY VALUE is "$KEY"'

If you just use Split() then you may not actually get a 'pair'
but rather a map with length >= 3.

mmcgrana commented:
Hmm interesting. Thanks for pointing this out @smoser. This does seem more correct to me, but also pretty subtle. @eliben wdyt?

smoser commented:
I agree it's a small thing. But my googling bright me to this example.  I
used it, and then later realized the bug in the copied code. I figured I'd
save others the trouble if finding it themselves.

Thanks for your work. I've found it useful many times.

On Wed, Oct 9, 2019, 7:24 PM Mark McGranaghan <notifications@github.com>
wrote:

> Hmm interesting. Thanks for pointing this out @smoser
> <https://github.com/smoser>. This does seem more correct to me, but also
> pretty subtle. @eliben <https://github.com/eliben> wdyt?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mmcgrana/gobyexample/pull/284?email_source=notifications&email_token=AAAETMPBTPTB4ULKAFYIAODQNZR4BA5CNFSM4I7EH7FKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEAZYNDI#issuecomment-540247693>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAETMINWY7DE7JM5P23QN3QNZR4BANCNFSM4I7EH7FA>
> .
>

eliben commented:
Technically this is correct if env values contain `=`s. But this sample currently doesn't show values at all, so it's not buggy, strictly speaking. For showing the key, it's correct.

However, this makes me think we're conflating multiple things in this example here. Perhaps we should separate concerns and not parse the env values at all, just showing how to access `os.Environ`, basically doing this:

```
	for _, e := range os.Environ() {
               fmt.Println(e)
	}
```

We already demonstrate `strings.Split` elsewhere (in String Functions).

@mmcgrana if you think keeping the parsing/splitting here is useful, then switching to `SplitN` SGTM, but then the comment should be updated too.

eliben commented:
> One option is definitely to not show access to the actual environment variables (as suggested in the `Println` change above).
> 
> I think that that is kind of missing the point though. No one (or _very_ few) is going to want to read environment variables without looking at key _and_ value. So why not show how to do that properly?

That's a fair point

eliben commented on examples/environment-variables/environment-variables.go:
```diff
@@ -27,7 +27,7 @@ func main() {
 	fmt.Println()
 	for _, e := range os.Environ() {
-		pair := strings.Split(e, "=")
+		pair := strings.SplitN(e, "=", 2)
```
You should update the comment above as well, as it mentions `strings.Split`

smoser commented:
Will do

On Thu, Oct 10, 2019, 5:01 PM Eli Bendersky <notifications@github.com>
wrote:

> *@eliben* requested changes on this pull request.
> ------------------------------
>
> In examples/environment-variables/environment-variables.go
> <https://github.com/mmcgrana/gobyexample/pull/284#discussion_r333733308>:
>
> > @@ -27,7 +27,7 @@ func main() {
>  	// get the key and value. Here we print all the keys.
>  	fmt.Println()
>  	for _, e := range os.Environ() {
> -		pair := strings.Split(e, "=")
> +		pair := strings.SplitN(e, "=", 2)
>
> You should update the comment above as well, as it mentions strings.Split
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mmcgrana/gobyexample/pull/284?email_source=notifications&email_token=AAAETMITDYIQYNTSPWDWEBLQN6JZFA5CNFSM4I7EH7FKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHTPONI#pullrequestreview-300349237>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAETMPARU7LQO4AWVBPFFDQN6JZFANCNFSM4I7EH7FA>
> .
>

mmcgrana commented:
Thanks for the thoughtful discussion guys. Here's why I net out:

* I'll change to `SplitN` in the code so that copy-paste does the right thing.
* I'll update the comment to reference `SplitN`, but I'm not going to digress to explain the subtlety in the comments, as we've already covered strings functions in an earlier example

smoser commented:
Good enough for me. Thanks

On Sun, Oct 13, 2019, 9:29 PM Mark McGranaghan <notifications@github.com>
wrote:

> Merged #284 <https://github.com/mmcgrana/gobyexample/pull/284> into
> master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mmcgrana/gobyexample/pull/284?email_source=notifications&email_token=AAAETMJXP77WBAZ7QAJ7SATQOPDOBA5CNFSM4I7EH7FKYY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGOUF4MFLI#event-2709045933>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAETMKHGKNNXNLA5DFJS63QOPDOBANCNFSM4I7EH7FA>
> .
>
